### PR TITLE
fix(customizer): Allow the user to select the uploaded output checkpoints

### PR DESCRIPTION
**Pull request by soluwalana**
## Summary

Training always published the lowest-validation-loss checkpoint when a `LOWEST_VAL` symlink existed, even if that checkpoint was from an earlier epoch. A 3-epoch embedding run could therefore upload a mid-run checkpoint and drop the last epoch.

`training.checkpoint_selection` now chooses what is published: `best` (default, lowest val loss), `last` (end of training), or `both` (best at the fileset root, last under `alternates/last`). Per-checkpoint val-loss stats are written into the fileset as `checkpoint_stats.json`.

## Changes

- Add `CheckpointSelection` (`best` | `last` | `both`) on the training job API and compile it into the training step config.
- Resolve `LOWEST_VAL` vs `LATEST` from that selection instead of always preferring lowest val.
- Process selected checkpoints by staging each under `output/{label}`, promoting the published one to the fileset root, and moving a leftover `last` under `alternates/last`.
- For embedding / cross-encoder jobs, convert each selected checkpoint (ONNX + HF) before the layout moves.
- Record val loss at each save and copy `checkpoint_stats.json` into the uploaded fileset.
- Cover selection, fileset layout, compiler plumbing, and stats writing in unit tests.

## Type of Change

- [x] Code change (feature, bug fix, or refactor)
- [ ] Code change with documentation updates
- [ ] Documentation only
- [ ] Contributor tooling or automation
- [ ] CI, build, or test infrastructure

## Quality Gates

- [x] Tests added or updated for changed behavior
- [ ] Existing tests cover changed behavior — justification:
- [ ] Tests not applicable — justification:
- [ ] Documentation updated for user-visible behavior
- [x] Documentation not applicable — justification: user-facing behavior is the new `training.checkpoint_selection` field and its schema description; there is no separate docs page for this Customizer training knob.

## Verification

- [x] Pull request title follows the repository's Conventional Commit format
- [x] Every commit includes an appropriate `Signed-off-by:` trailer
- [ ] `uv run pre-commit run -a` passes, or any blocked checks are identified below
- [x] Targeted tests pass, or tests are marked not applicable above
- [x] No secrets, API keys, or credentials are included

Targeted validation:

- `pytest services/automodel/tests`: 207 passed, 9 skipped.
- `ty check` on the non-excluded touched files: all checks passed. `tasks/training/backends/` remains in `[tool.ty.src].exclude` (GPU-container imports).
- `ruff check` / `ruff format --check` on automodel src and tests: passed.
- `uv run pre-commit run -a` was not run for this description update.

## Summary by CodeRabbit

* **New Features**
  * Choose to publish the best checkpoint, final checkpoint, or both; when publishing both, the final checkpoint is available under `alternates/last`.
  * Enable distributed in-batch negatives for bi-encoder retrieval training.
  * Training records checkpoint statistics, including steps, losses, and best-metric information.
  * Improved encoder and ONNX output compatibility.

* **Documentation**
  * Updated CLI documentation for checkpoint selection and distributed in-batch negatives.
  * Documented fields that must be provided through `--spec` or `--spec-file`.

### CODE REVIEW

File: services/automodel/src/nmp/automodel/tasks/training/backends/checkpoints.py [hunk 1]
```diff
@@ -794,3 +896,60 @@ def process_checkpoint(
...
+    conversion and sanitizing happen identically for each. The published
+    checkpoint is then moved up to the root, and a ``last`` that was not the
+    published one moves under ``alternates/``.
+    """
+    infos: dict[str, CheckpointInfo] = {}
+    for label, checkpoint_path in checkpoints.items():
+        infos[label] = process_checkpoint(
+            checkpoint_path,
+            output_path / label,
+            customizer_config,
```
[the remark this answers is not in the document]
Comment (soluwalana) [reply services/automodel/src/nmp/automodel/tasks/training/backends/checkpoints.py:938]: Going to skip this fix. 

Technically true, but in practice irrelevant.

**Review by gabwow — APPROVED** [review]

File: services/automodel/src/nmp/automodel/tasks/training/backends/checkpoints.py [hunk 2]
```diff
@@ -775,13 +871,19 @@ def process_checkpoint(
 
     if (is_embedding or is_cross_encoder) and checkpoint_format != CheckpointFormat.HF_PEFT:
         export_cfg = _resolve_export_config(customizer_config)
+        # Full-SFT copy still has Automodel auto_map until sanitize runs.
+        load_trust_remote_code = True if finetuning_type == FinetuningType.ALL_WEIGHTS else job_trust_remote_code
```
Comment (anubhutivyas) [inline services/automodel/src/nmp/automodel/tasks/training/backends/checkpoints.py:875]: why do we need `trust_remote_code` for all `all_weights` cross_encode and bi_encoder models?

Comment (soluwalana) [reply services/automodel/src/nmp/automodel/tasks/training/backends/checkpoints.py:875]: Because the onnx module needs to read the custom heads and pooling layers that get generated by automodel for appropriate graph tracing. The output model doesn't get the trust_remote_code == True forced flag

File: services/automodel/src/nmp/automodel/tasks/training/backends/finetune.py [hunk 3]
```diff
@@ -31,6 +34,23 @@
 
 logger = logging.getLogger(__name__)
 
+CHECKPOINT_STATS_FILENAME = "checkpoint_stats.json"
```
Comment (anubhutivyas) [inline services/automodel/src/nmp/automodel/tasks/training/backends/finetune.py:37]: nit: may be define it just once somewhere common?

File: docs/cli/reference.mdx [hunk 4]
```diff
@@ -9656,7 +9657,7 @@ nemo customization automodel.jobs submit [OPTIONS]
...
+Some spec fields cannot be represented as CLI flags: `training.training_type`, `training.recipe`, `training.finetuning_type`, `training.lora.target_modules`, `training.lora.exclude_modules`, `training.precision`, `training.attn_implementation`, `training.teacher_precision`, `training.retrieval.export.primary`, `training.retrieval.export.precision`, `training.retrieval.export.attn_implementation`, `training.retrieval.export.pooling`, `schedule.checkpoint_selection`, `schedule.progress_reporting.time_series_metrics`, `optimizer.optimizer`, `optimizer.lr_decay_style`, `integrations.wandb.tags`, `integrations.mlflow.tags`. Supply them with --spec or --spec-file. Individual flags override values from the supplied spec.
```
Comment (anubhutivyas) [inline docs/cli/reference.mdx:9661]: `training.checkpoint_selection.`?

**Review by anubhutivyas — APPROVED** [review]
super nit comments, but overall LGTM

File: docs/cli/reference.mdx (hunk 4 above)
Comment (soluwalana) [reply docs/cli/reference.mdx:9661]: Yeah, it's a new parameter. It allows us to toggle "last" checkpoint "best" checkpoint, or "both" in the fileset

File: services/automodel/src/nmp/automodel/tasks/training/backends/finetune.py (hunk 3 above)
Comment (soluwalana) [reply services/automodel/src/nmp/automodel/tasks/training/backends/finetune.py:37]: I'm only using it in this file currently, I can move it later if I reuse it somewhere else
